# Gather (Aggregate) Projects Data
2024-01-03 ZD  


In [1]:
# Method to import from parent directory
import os
import sys
root_dir = os.path.abspath(os.path.join(os.getcwd(), "../"))
sys.path.append(root_dir)
import config
from datetime import datetime

import pandas as pd


---TIMESTAMP OVERRIDE IN USE---
---Performing action using 2024-01-04 instead of current timestamp.---
---Change the OVERRIDE_DATE in config.py to None to restore default behavior.---




In [3]:
# Load Grants data
grant_filepath = '../'+config.GRANTS_INTERMED_PATH
grants_df = pd.read_csv(grant_filepath)
grants_df

grant_id queried_project_id  application_id  fiscal_year  \
0       1U24CA233218-01        U24CA233218         9627071         2018   
1       1UG3CA233229-01        UG3CA233229         9627196         2018   
2       1UG3CA233251-01        UG3CA233251         9627357         2018   
3       1UG3CA233282-01        UG3CA233282         9627516         2018   
4     1UG3CA233314-01A1        UG3CA233314         9870730         2019   
...                 ...                ...             ...          ...   
2932    5U24CA230144-02        U24CA230144         9786693         2019   
2933    5U24CA230144-03        U24CA230144        10006517         2020   
2934    5U24CA230144-04        U24CA230144        10249162         2021   
2935    5U24CA230144-06        U24CA230144        10476535         2022   
2936    6U24CA230144-05        U24CA230144        10601411         2021   

                                          project_title  \
0     Accelerating Colorectal Cancer Screening and F...   
1                                       ACCISIS-Chicago   
2     Scaling Colorectal Cancer Screening Through Ou...   
3     (ACCSIS) Accelerating Colorectal Cancer Screen...   
4     CRC-HUB-SPOKE: A ColoRectal Cancer screening H...   
...                                                 ...   
2932  Consortium on Translational Research in Early ...   
2933  Consortium on Translational Research in Early ...   
2934  Consortium on Translational Research in Early ...   
2935  Consortium on Translational Research in Early ...   
2936  Consortium on Translational Research in Early ...   

                                          abstract_text  \
0     PROJECT SUMMARY/ABSTRACT To advance research i...   
1     PROJECT SUMMARY Screening for colorectal cance...   
2     PROJECT SUMMARY Despite strong evidence that c...   
3     Accelerating Colorectal Cancer Screening throu...   
4     ABSTRACT Colorectal cancer (CRC) screening and...   
...                                                 ...   
2932  Project Summary The key for the success of the...   
2933  Project Summary The key for the success of the...   
2934  Project Summary The key for the success of the...   
2935  Project Summary The key for the success of the...   
2936  Project Summary The key for the success of the...   

                                               keywords  \
0     Adoption;Award;Behavior Therapy;Behavioral;Bio...   
1     Adoption;African American;Area;Asian Americans...   
2     Adoption;African American;American Indians;App...   
3     Address;Administrator;Advisory Committees;Age;...   
4     Address;Adoption;Advisory Committees;Californi...   
...                                                 ...   
2932  Bioinformatics;Biological Markers;Biometry;Can...   
2933  Bioinformatics;Biological Markers;Biological S...   
2934  Bioinformatics;Biological Markers;Biological S...   
2935  Bioinformatics;Biological Markers;Biological S...   
2936                                                NaN   

                                principal_investigators    program_officers  \
0                         Sujha Subramanian; Sylvia Tan  Genevieve M Grimes   
1                          Karen E Kim; Blase N. Polite  Genevieve M Grimes   
2                                     Daniel S. Reuland  Genevieve M Grimes   
3                     Mark B Dignan; Electra D. Paskett  Genevieve M Grimes   
4     Sheila Faye Castaneda; Samir Gupta; Maria Elen...  Genevieve M Grimes   
...                                                 ...                 ...   
2932                                        Ziding Feng   Guillermo Marquez   
2933                                        Ziding Feng   Guillermo Marquez   
2934                                        Ziding Feng   Guillermo Marquez   
2935                                        Ziding Feng   Guillermo Marquez   
2936                                        Ziding Feng   Guillermo Marquez   

      award_amount  ...     award_noti

## First Try

In [20]:
# Convert award notice data to datetime format
grants_df['award_notice_date'] = pd.to_datetime(grants_df['award_notice_date'])

In [21]:
grants_df

grant_id queried_project_id  application_id  fiscal_year  \
0       1U24CA233218-01        U24CA233218         9627071         2018   
1       1UG3CA233229-01        UG3CA233229         9627196         2018   
2       1UG3CA233251-01        UG3CA233251         9627357         2018   
3       1UG3CA233282-01        UG3CA233282         9627516         2018   
4     1UG3CA233314-01A1        UG3CA233314         9870730         2019   
...                 ...                ...             ...          ...   
6647    5U24CA230144-02        U24CA230144         9786693         2019   
6648    5U24CA230144-03        U24CA230144        10006517         2020   
6649    5U24CA230144-04        U24CA230144        10249162         2021   
6650    5U24CA230144-06        U24CA230144        10476535         2022   
6651    6U24CA230144-05        U24CA230144        10601411         2021   

                                          project_title  \
0     Accelerating Colorectal Cancer Screening and F...   
1                                       ACCISIS-Chicago   
2     Scaling Colorectal Cancer Screening Through Ou...   
3     (ACCSIS) Accelerating Colorectal Cancer Screen...   
4     CRC-HUB-SPOKE: A ColoRectal Cancer screening H...   
...                                                 ...   
6647  Consortium on Translational Research in Early ...   
6648  Consortium on Translational Research in Early ...   
6649  Consortium on Translational Research in Early ...   
6650  Consortium on Translational Research in Early ...   
6651  Consortium on Translational Research in Early ...   

                                          abstract_text  \
0     PROJECT SUMMARY/ABSTRACT To advance research i...   
1     PROJECT SUMMARY Screening for colorectal cance...   
2     PROJECT SUMMARY Despite strong evidence that c...   
3     Accelerating Colorectal Cancer Screening throu...   
4     ABSTRACT Colorectal cancer (CRC) screening and...   
...                                                 ...   
6647  Project Summary The key for the success of the...   
6648  Project Summary The key for the success of the...   
6649  Project Summary The key for the success of the...   
6650  Project Summary The key for the success of the...   
6651  Project Summary The key for the success of the...   

                                               keywords  \
0     Adoption;Award;Behavior Therapy;Behavioral;Bio...   
1     Adoption;African American;Area;Asian Americans...   
2     Adoption;African American;American Indians;App...   
3     Address;Administrator;Advisory Committees;Age;...   
4     Address;Adoption;Advisory Committees;Californi...   
...                                                 ...   
6647  Bioinformatics;Biological Markers;Biometry;Can...   
6648  Bioinformatics;Biological Markers;Biological S...   
6649  Bioinformatics;Biological Markers;Biological S...   
6650  Bioinformatics;Biological Markers;Biological S...   
6651                                                NaN   

                                principal_investigators    program_officers  \
0                         Sujha Subramanian; Sylvia Tan  Genevieve M Grimes   
1                          Karen E Kim; Blase N. Polite  Genevieve M Grimes   
2                                     Daniel S. Reuland  Genevieve M Grimes   
3                     Mark B Dignan; Electra D. Paskett  Genevieve M Grimes   
4     Sheila Faye Castaneda; Samir Gupta; Maria Elen...  Genevieve M Grimes   
...                                                 ...                 ...   
6647                                        Ziding Feng   Guillermo Marquez   
6648                                        Ziding Feng   Guillermo Marquez   
6649                                        Ziding Feng   Guillermo Marquez   
6650                                        Ziding Feng   Guillermo Marquez   
6651                                        Ziding Feng   Guillermo Marquez   

      award_amount  ...         award_

In [22]:
# Create a new DataFrame for projects
projects_df = pd.DataFrame()

# Aggregate grants data by queried_project_id
projects_df['project_id'] = grants_df['queried_project_id'].unique()


In [23]:
projects_df

project_id
0     U24CA233218
1     UG3CA233229
2     UG3CA233251
3     UG3CA233282
4     UG3CA233314
...           ...
2616  U01CA230705
2617  U01CA230997
2618  U01CA283931
2619  U01CA283935
2620  U24CA230144

[2621 rows x 1 columns]

In [59]:
# Columns to aggregate with corresponding aggregation type
aggregation_columns = {
    'project_title': 'max',
    'abstract_text': 'max',
    'project_start_date': 'min',
    'project_end_date': 'max',
    'opportunity_number': 'first',
    'api_source_search': 'first',
    'org_name': 'first',
    'org_city': 'first',
    'org_state': 'first',
    'org_country': 'first',
    'program.program_id': 'first'
}


In [61]:
for column_name in aggregation_columns.keys():
    print(column_name in grants_df.columns)

True
True
True
True
True
True
True
True
True
True
True


In [92]:
test_df = grants_df[(grants_df['queried_project_id'] == 'U24CA233218') |
                    (grants_df['queried_project_id'] == 'UG3CA233229')
                    ]
test_df

grant_id queried_project_id  application_id  fiscal_year  \
0    1U24CA233218-01        U24CA233218         9627071         2018   
1    1UG3CA233229-01        UG3CA233229         9627196         2018   
6  3U24CA233218-01S1        U24CA233218        10006947         2019   

                                       project_title  \
0  Accelerating Colorectal Cancer Screening and F...   
1                                    ACCISIS-Chicago   
6  Accelerating Colorectal Cancer Screening and F...   

                                       abstract_text  \
0  PROJECT SUMMARY/ABSTRACT To advance research i...   
1  PROJECT SUMMARY Screening for colorectal cance...   
6  PROJECT SUMMARY/ABSTRACT To advance research i...   

                                            keywords  \
0  Adoption;Award;Behavior Therapy;Behavioral;Bio...   
1  Adoption;African American;Area;Asian Americans...   
6  Adoption;Award;Behavior Therapy;Behavioral;Bio...   

         principal_investigators    program_officers  award_amount  ...  \
0  Sujha Subramanian; Sylvia Tan  Genevieve M Grimes       2863779  ...   
1   Karen E Kim; Blase N. Polite  Genevieve M Grimes        810000  ...   
6              Sujha Subramanian  Genevieve M Grimes        785056  ...   

          award_notice_date    project_start_date      project_end_date  \
0 2018-09-20 12:09:00+00:00  2018-09-30T12:09:00Z  2024-08-31T12:08:00Z   
1 2018-09-21 12:09:00+00:00  2018-09-21T12:09:00Z  2019-08-31T12:08:00Z   
6 2019-09-20 12:09:00+00:00  2018-09-30T12:09:00Z  2023-08-31T12:08:00Z   

  opportunity_number api_source_search                     org_name  \
0      RFA-CA-17-039     nofo_CA17-039  RESEARCH TRIANGLE INSTITUTE   
1      RFA-CA-17-038    nofo_CA-17-038        UNIVERSITY OF CHICAGO   
6      RFA-CA-17-039     nofo_CA17-039  RESEARCH TRIANGLE INSTITUTE   

                 org_city org_state    org_country program.program_id  
0  RESEARCH TRIANGLE PARK        NC  UNITED STATES             accsis  
1                 CHICAGO        IL  UNITED STATES             accsis  
6  RESEARCH TRIANGLE PARK        NC  UNITED STATES             accsis  

[3 rows x 21 columns]

In [93]:
test_agg_df = pd.DataFrame()

for column_name in aggregation_columns.keys():

    test_agg_df[column_name] = test_df.groupby('queried_project_id')[column_name].agg('last')

test_agg_df


project_title  \
queried_project_id                                                      
U24CA233218         Accelerating Colorectal Cancer Screening and F...   
UG3CA233229                                           ACCISIS-Chicago   

                                                        abstract_text  \
queried_project_id                                                      
U24CA233218         PROJECT SUMMARY/ABSTRACT To advance research i...   
UG3CA233229         PROJECT SUMMARY Screening for colorectal cance...   

                      project_start_date      project_end_date  \
queried_project_id                                               
U24CA233218         2018-09-30T12:09:00Z  2023-08-31T12:08:00Z   
UG3CA233229         2018-09-21T12:09:00Z  2019-08-31T12:08:00Z   

                   opportunity_number api_source_search  \
queried_project_id                                        
U24CA233218             RFA-CA-17-039     nofo_CA17-039   
UG3CA233229             RFA-CA-17-038    nofo_CA-17-038   

                                       org_name                org_city  \
queried_project_id                                                        
U24CA233218         RESEARCH TRIANGLE INSTITUTE  RESEARCH TRIANGLE PARK   
UG3CA233229               UNIVERSITY OF CHICAGO                 CHICAGO   

                   org_state    org_country program.program_id  
queried_project_id                                              
U24CA233218               NC  UNITED STATES             accsis  
UG3CA233229               IL  UNITED STATES             accsis

In [94]:
test_agg_df = pd.DataFrame()

for column_name, agg_type in aggregation_columns.items():

    test_agg_df[column_name] = test_df.groupby('queried_project_id')[column_name].agg(agg_type)

test_agg_df

project_title  \
queried_project_id                                                      
U24CA233218         Accelerating Colorectal Cancer Screening and F...   
UG3CA233229                                           ACCISIS-Chicago   

                                                        abstract_text  \
queried_project_id                                                      
U24CA233218         PROJECT SUMMARY/ABSTRACT To advance research i...   
UG3CA233229         PROJECT SUMMARY Screening for colorectal cance...   

                      project_start_date      project_end_date  \
queried_project_id                                               
U24CA233218         2018-09-30T12:09:00Z  2024-08-31T12:08:00Z   
UG3CA233229         2018-09-21T12:09:00Z  2019-08-31T12:08:00Z   

                   opportunity_number api_source_search  \
queried_project_id                                        
U24CA233218             RFA-CA-17-039     nofo_CA17-039   
UG3CA233229             RFA-CA-17-038    nofo_CA-17-038   

                                       org_name                org_city  \
queried_project_id                                                        
U24CA233218         RESEARCH TRIANGLE INSTITUTE  RESEARCH TRIANGLE PARK   
UG3CA233229               UNIVERSITY OF CHICAGO                 CHICAGO   

                   org_state    org_country program.program_id  
queried_project_id                                              
U24CA233218               NC  UNITED STATES             accsis  
UG3CA233229               IL  UNITED STATES             accsis

In [95]:
test_df.groupby('queried_project_id')['fiscal_year'].first()

queried_project_id
U24CA233218    2018
UG3CA233229    2018
Name: fiscal_year, dtype: int64

In [96]:
def get_identical_values(grants_df, column_name):
    values = grants_df.groupby('queried_project_id')[column_name].agg('first')
    # Validation check for identical values
    if not values.nunique() == 1:
        raise ValueError(f"Validation failed: {column_name} does not have identical values for all rows with the same project_id.")
    return values

def get_newest_values(grants_df, column_name):
    # Find the index of the row with the maximum 'award_notice_date'
    idxmax_date = grants_df.groupby('queried_project_id')['award_notice_date'].idxmax()
    # Extract values based on the found index
    return grants_df.loc[idxmax_date, column_name]

def get_oldest_values(grants_df, column_name):
    # Find the index of the row with the minimum 'award_notice_date'
    idxmin_date = grants_df.groupby('queried_project_id')['award_notice_date'].idxmin()
    # Extract values based on the found index
    return grants_df.loc[idxmin_date, column_name]

In [97]:
def gather_project_data(grants_df):
    # Create a new DataFrame for projects
    projects_df = pd.DataFrame()

    # Columns to aggregate with corresponding aggregation type
    aggregation_columns = {
        'project_title': 'newest',
        'abstract_text': 'newest',
        'project_start_date': 'oldest',
        'project_end_date': 'newest',
        'opportunity_number': 'identical',
        'api_source_search': 'identical',
        'org_name': 'identical',
        'org_city': 'identical',
        'org_state': 'identical',
        'org_country': 'identical',
        'program.program_id': 'identical'
    }

    # Apply aggregation for each column
    for column, aggregation_type in aggregation_columns.items():
        if column not in grants_df:
            raise ValueError(f"Expected column not found in grants data: {column}.")
        else:
            if aggregation_type == 'identical':
                projects_df[column] = get_identical_values(grants_df, column)
            elif aggregation_type == 'newest':
                projects_df[column] = get_newest_values(grants_df, column)
            elif aggregation_type == 'oldest':
                projects_df[column] = get_oldest_values(grants_df, column)
    
    return projects_df

In [98]:
test_df

grant_id queried_project_id  application_id  fiscal_year  \
0    1U24CA233218-01        U24CA233218         9627071         2018   
1    1UG3CA233229-01        UG3CA233229         9627196         2018   
6  3U24CA233218-01S1        U24CA233218        10006947         2019   

                                       project_title  \
0  Accelerating Colorectal Cancer Screening and F...   
1                                    ACCISIS-Chicago   
6  Accelerating Colorectal Cancer Screening and F...   

                                       abstract_text  \
0  PROJECT SUMMARY/ABSTRACT To advance research i...   
1  PROJECT SUMMARY Screening for colorectal cance...   
6  PROJECT SUMMARY/ABSTRACT To advance research i...   

                                            keywords  \
0  Adoption;Award;Behavior Therapy;Behavioral;Bio...   
1  Adoption;African American;Area;Asian Americans...   
6  Adoption;Award;Behavior Therapy;Behavioral;Bio...   

         principal_investigators    program_officers  award_amount  ...  \
0  Sujha Subramanian; Sylvia Tan  Genevieve M Grimes       2863779  ...   
1   Karen E Kim; Blase N. Polite  Genevieve M Grimes        810000  ...   
6              Sujha Subramanian  Genevieve M Grimes        785056  ...   

          award_notice_date    project_start_date      project_end_date  \
0 2018-09-20 12:09:00+00:00  2018-09-30T12:09:00Z  2024-08-31T12:08:00Z   
1 2018-09-21 12:09:00+00:00  2018-09-21T12:09:00Z  2019-08-31T12:08:00Z   
6 2019-09-20 12:09:00+00:00  2018-09-30T12:09:00Z  2023-08-31T12:08:00Z   

  opportunity_number api_source_search                     org_name  \
0      RFA-CA-17-039     nofo_CA17-039  RESEARCH TRIANGLE INSTITUTE   
1      RFA-CA-17-038    nofo_CA-17-038        UNIVERSITY OF CHICAGO   
6      RFA-CA-17-039     nofo_CA17-039  RESEARCH TRIANGLE INSTITUTE   

                 org_city org_state    org_country program.program_id  
0  RESEARCH TRIANGLE PARK        NC  UNITED STATES             accsis  
1                 CHICAGO        IL  UNITED STATES             accsis  
6  RESEARCH TRIANGLE PARK        NC  UNITED STATES             accsis  

[3 rows x 21 columns]

In [99]:
gather_project_data(test_df)

ValueError: Validation failed: opportunity_number does not have identical values for all rows with the same project_id.

In [101]:
test_df['opportunity_number']

0    RFA-CA-17-039
1    RFA-CA-17-038
6    RFA-CA-17-039
Name: opportunity_number, dtype: object

In [89]:
get_newest_values(test_df, 'project_start_date')

6    2018-09-30T12:09:00Z
Name: project_start_date, dtype: object

## Start over and try again

In [107]:
def validate_identical_values(df, field_name):
    inconsistent_rows = df[df.groupby('queried_project_id')[field_name].transform('nunique') > 1]
    for index, row in inconsistent_rows.iterrows():
        print(f"NOTICE: Unexpected mismatch in {field_name} for {row['queried_project_id']}")

def get_newest_or_oldest_value(df, field_name, newest=True):
    return df.sort_values('award_notice_date', ascending=not newest).groupby('queried_project_id')[field_name].first()

def gather_project_data(grants_df):
    """Aggregates grants data into project-level information, handling inconsistencies and validation."""

    field_aggregation_map = {
        'queried_project_id': 'Identical',
        'project_title': 'Newest',
        'abstract_text': 'Newest',
        'project_start_date': 'Oldest',
        'project_end_date': 'Newest',
        'opportunity_number': 'Identical',
        'api_source_search': 'Identical',
        'org_name': 'Identical',
        'org_city': 'Identical',
        'org_state': 'Identical',
        'org_country': 'Identical',
        'program.program_id': 'Identical'
    }

    projects_df = pd.DataFrame(columns=field_aggregation_map.keys())

    # Validate and aggregate "Identical" fields
    for field_name, aggregation_type in field_aggregation_map.items():
        if aggregation_type == 'Identical':
            validate_identical_values(grants_df, field_name)
            projects_df[field_name] = grants_df.groupby('queried_project_id')[field_name].first()

    # Get "Newest" or "Oldest" fields
    for field_name, aggregation_type in field_aggregation_map.items():
        if aggregation_type in ('Newest', 'Oldest'):
            projects_df[field_name] = get_newest_or_oldest_value(grants_df, field_name, newest=aggregation_type == 'Newest')

    # Rename 'queried_project_id' to 'project_id'
    projects_df.rename(columns={'queried_project_id': 'project_id'}, inplace=True)

    return projects_df

In [108]:
gather_project_data(test_df)

project_id  \
queried_project_id                
U24CA233218         U24CA233218   
UG3CA233229         UG3CA233229   

                                                        project_title  \
queried_project_id                                                      
U24CA233218         Accelerating Colorectal Cancer Screening and F...   
UG3CA233229                                           ACCISIS-Chicago   

                                                        abstract_text  \
queried_project_id                                                      
U24CA233218         PROJECT SUMMARY/ABSTRACT To advance research i...   
UG3CA233229         PROJECT SUMMARY Screening for colorectal cance...   

                      project_start_date      project_end_date  \
queried_project_id                                               
U24CA233218         2018-09-30T12:09:00Z  2023-08-31T12:08:00Z   
UG3CA233229         2018-09-21T12:09:00Z  2019-08-31T12:08:00Z   

                   opportunity_number api_source_search  \
queried_project_id                                        
U24CA233218             RFA-CA-17-039     nofo_CA17-039   
UG3CA233229             RFA-CA-17-038    nofo_CA-17-038   

                                       org_name                org_city  \
queried_project_id                                                        
U24CA233218         RESEARCH TRIANGLE INSTITUTE  RESEARCH TRIANGLE PARK   
UG3CA233229               UNIVERSITY OF CHICAGO                 CHICAGO   

                   org_state    org_country program.program_id  
queried_project_id                                              
U24CA233218               NC  UNITED STATES             accsis  
UG3CA233229               IL  UNITED STATES             accsis

In [109]:
gather_project_data(grants_df)

NOTICE: Unexpected mismatch in opportunity_number for UH3CA244298
NOTICE: Unexpected mismatch in opportunity_number for UH3CA244298
NOTICE: Unexpected mismatch in opportunity_number for UH3CA244298
NOTICE: Unexpected mismatch in opportunity_number for UH3CA244298
NOTICE: Unexpected mismatch in opportunity_number for R01CA239701
NOTICE: Unexpected mismatch in opportunity_number for R01CA239701
NOTICE: Unexpected mismatch in opportunity_number for R01CA239701
NOTICE: Unexpected mismatch in opportunity_number for R01CA239701
NOTICE: Unexpected mismatch in opportunity_number for U10CA180821
NOTICE: Unexpected mismatch in opportunity_number for U10CA180821
NOTICE: Unexpected mismatch in opportunity_number for U10CA180821
NOTICE: Unexpected mismatch in opportunity_number for U10CA180821
NOTICE: Unexpected mismatch in opportunity_number for U10CA180821
NOTICE: Unexpected mismatch in opportunity_number for U10CA180821
NOTICE: Unexpected mismatch in opportunity_number for U10CA180821
NOTICE: Un

project_id  \
queried_project_id                
26114002B91         26114002B91   
26120080000         26120080000   
26120150000         26120150000   
D43TW009340         D43TW009340   
D43TW009343         D43TW009343   
...                         ...   
UM1CA228823         UM1CA228823   
UM1CA233033         UM1CA233033   
UM1CA233035         UM1CA233035   
UM1CA233080         UM1CA233080   
UM1HG006348         UM1HG006348   

                                                        project_title  \
queried_project_id                                                      
26114002B91         Development and Expansion of Digital Tools to ...   
26120080000         Burkitt Lymphoma Genome Sequencing Project (BL...   
26120150000         HIV and Tumor Molecular Characterization Proje...   
D43TW009340         UJMT FGHF Launching Future Leaders in Global H...   
D43TW009343         University of California Launching Future Lead...   
...                                                               ...   
UM1CA228823         Children's Oncology Group Pediatric Early Phas...   
UM1CA233033         Enhanced, EHR-facilitated Cancer Symptom Contr...   
UM1CA233035         Implementation and Evaluation of an Expanded B...   
UM1CA233080         SIMPRO Research Center: Integration and Implem...   
UM1HG006348         Consortium for large-scale production and phen...   

                                                        abstract_text  \
queried_project_id                                                      
26114002B91         Earlier this year, NCI launched the Cancer Cen...   
26120080000         Burkitt Lymphoma (BL) patients would benefit f...   
26120150000         HIV and Tumor Molecular Characterization Proje...   
D43TW009340         PROJECT SUMMARY In this supplemental applicati...   
D43TW009343         In response to RFA-TW-21-004, the University o...   
...                                                               ...   
UM1CA228823         The mission of the Pediatric Early Phase Clini...   
UM1CA233033         Cancer and its treatment engender disabling sy...   
UM1CA233035         Cancer, and cancer treatment, cause many sympt...   
UM1CA233080         Project Summary In 2018, more than 1.7 million...   
UM1HG006348         ABSTRACT The long-term goal of the Internation...   

                      project_start_date      project_end_date  \
queried_project_id                                               
26114002B91         2019-09-15T12:09:00Z  2020-09-14T12:09:00Z   
26120080000         2008-09-26T12:09:00Z  2018-09-25T12:09:00Z   
26120150000         2018-09-24T12:09:00Z  2022-03-31T12:03:00Z   
D43TW009340         2012-04-04T12:04:00Z  2027-06-30T12:06:00Z   
D43TW009343         2023-07-04T12:07:00Z  2024-06-30T12:06:00Z   
...                                  ...                   ...   
UM1CA228823         2018-09-20T12:09:00Z  2024-07-31T12:07:00Z   
UM1CA233033         2018-09-21T12:09:00Z  2024-08-31T12:08:00Z   
UM1CA233035         2018-09-24T12:09:00Z  2024-08-31T12:08:00Z   
UM1CA233080         2018-09-20T12:09:00Z  2024-08-31T12:08:00Z   
UM1HG006348         2011-09-28T12:09:00Z  2022-08-31T12:08:00Z   

                   opportunity_number                 api_source_search  \
queried_project_id                                                        
26114002B91                      None  award_26114002B91019F00132-0-0-1   
26120080000                      None      award_261200800001E-52-0-113   
26120150000                      None  award_261201500003I-0-26100065-1   
D43TW009340                 PA-20-272                     nofo_PA20-272   
D43TW009343                 PA-20-272                     nofo_PA20-272   
...                               ...                               ...   
UM1CA228823                 PA-20-272                     nofo_PA20-272   
UM1CA233033             RFA-CA-17-042                    nofo_CA-17-042   
UM1CA233035             RFA-CA-17-042                   

In [113]:
grants_df[grants_df['queried_project_id'] == 'U10CA180821'].groupby('opportunity_number').size().reset_index()

opportunity_number   0
0          PA-20-272  14
1      RFA-CA-12-010  16
2      RFA-CA-17-056  11

## Try again
2024-01-05  
Use a new aggregation method to list out all unique opportunity numbers for a project.

#### Validate that date logic is working correctly

In [4]:
field_aggregation_map = {
    'queried_project_id': 'identical',
    'project_title': 'newest',
    'abstract_text': 'newest',
    'project_start_date': 'oldest',
    'project_end_date': 'newest',
    'opportunity_number': 'List',
    'api_source_search': 'identical',
    'org_name': 'identical',
    'org_city': 'identical',
    'org_state': 'identical',
    'org_country': 'identical',
    'program.program_id': 'identical'
}

In [7]:
# Fing projects with many grants
grants_df.groupby('queried_project_id').size().reset_index().sort_values(by=0, ascending=False)

queried_project_id   0
19         P50CA070907  37
381        U10CA180821  34
17         P50CA058223  32
18         P50CA062924  31
380        U10CA031946  29
..                 ...  ..
379        U01DE029255   1
437        U24CA224316   1
445        U24CA232979   1
446        U24CA232980   1
692        UM1CA233080   1

[693 rows x 2 columns]

In [11]:
test_df = grants_df[grants_df['queried_project_id'] == 'P50CA070907']

In [14]:
test_df.nunique()

grant_id                   37
queried_project_id          1
application_id             37
fiscal_year                24
project_title               5
abstract_text               7
keywords                   15
principal_investigators     4
program_officers            3
award_amount               33
nci_funded_amount          33
award_notice_date          36
project_start_date          3
project_end_date            9
opportunity_number          5
api_source_search           1
org_name                    1
org_city                    1
org_state                   1
org_country                 1
program.program_id          1
dtype: int64

In [26]:
# Get oldest and newest dates
oldest = sorted(test_df.award_notice_date.unique().tolist())[0]
newest = sorted(test_df.award_notice_date.unique().tolist())[-1]

print(f"Oldest data: {oldest}\n"
      f"Newest date: {newest}")

Oldest data: 2000-08-23T12:08:00Z
Newest date: 2023-08-21T12:08:00Z


In [27]:
# Title, abstract, and project end date should be pulled from the newest grant
test_df[test_df['award_notice_date'] == newest]

grant_id queried_project_id  application_id  fiscal_year  \
2825  3P50CA070907-24S1        P50CA070907        10845884         2023   
2841    5P50CA070907-24        P50CA070907        10701005         2023   

                              project_title  \
2825  Targeting Lung Cancer Vulnerabilities   
2841  Targeting Lung Cancer Vulnerabilities   

                                          abstract_text  \
2825  Overall SPORE Summary/Abstract Targeting Lung ...   
2841  Overall SPORE Summary/Abstract Targeting Lung ...   

                                               keywords  \
2825  Advocate;Biological Markers;Biological Models;...   
2841  Advocate;Biological Markers;Biological Models;...   

             principal_investigators program_officers  award_amount  ...  \
2825  John V. Heymach; John D. Minna     Peter Ujhazy        118034  ...   
2841  John V. Heymach; John D. Minna     Peter Ujhazy       2110656  ...   

         award_notice_date    project_start_date      project_end_date  \
2825  2023-08-21T12:08:00Z  1997-09-05T12:09:00Z  2025-08-31T12:08:00Z   
2841  2023-08-21T12:08:00Z  1997-09-05T12:09:00Z  2025-08-31T12:08:00Z   

     opportunity_number  api_source_search                        org_name  \
2825         PAR-18-313  award_P50CA070907  UT SOUTHWESTERN MEDICAL CENTER   
2841         PAR-18-313  award_P50CA070907  UT SOUTHWESTERN MEDICAL CENTER   

     org_city org_state    org_country       program.program_id  
2825   DALLAS        TX  UNITED STATES  spore_lung_cancer_texas  
2841   DALLAS        TX  UNITED STATES  spore_lung_cancer_texas  

[2 rows x 21 columns]

In [28]:
# Only project start date should be pulled from the oldest grant
test_df[test_df['award_notice_date'] == oldest]

grant_id queried_project_id  application_id  fiscal_year  \
2826  5P50CA070907-05        P50CA070907         6173222         2000   

                                 project_title  \
2826  UNIVERSITY OF TEXAS SPORE IN LUNG CANCER   

                                          abstract_text  \
2826  The overarching strategic plan of this SPORE i...   

                                               keywords  \
2826  genetic markers; lung neoplasms; neoplasm /can...   

     principal_investigators program_officers  award_amount  ...  \
2826           John D. Minna     Peter Ujhazy       2224767  ...   

         award_notice_date    project_start_date      project_end_date  \
2826  2000-08-23T12:08:00Z  1996-09-30T12:09:00Z  2003-04-30T12:04:00Z   

     opportunity_number  api_source_search                        org_name  \
2826                NaN  award_P50CA070907  UT SOUTHWESTERN MEDICAL CENTER   

     org_city org_state    org_country       program.program_id  
2826   DALLAS        TX  UNITED STATES  spore_lung_cancer_texas  

[1 rows x 21 columns]

In [108]:
def get_newest_or_oldest_value(df, field_name, date_field='award_notice_date', newest=True):
    """Returns the newest or oldest value of a field based on 'award_notice_date'."""

    value = (df.sort_values(date_field, ascending=not newest)
             .groupby(['queried_project_id'])
             [field_name].first())

    return value

In [32]:
# Build empty projects df
projects_df = pd.DataFrame(columns=field_aggregation_map.keys())

# Validate and aggregate fields based on aggregation type
for field_name, aggregation_type in field_aggregation_map.items():

    if aggregation_type in ('newest', 'oldest'):

        projects_df[field_name] = get_newest_or_oldest_value(test_df, field_name, newest=aggregation_type == 'newest')

projects_df

queried_project_id                          project_title  \
queried_project_id                                                             
P50CA070907                       NaN  Targeting Lung Cancer Vulnerabilities   

                                                        abstract_text  \
queried_project_id                                                      
P50CA070907         Overall SPORE Summary/Abstract Targeting Lung ...   

                      project_start_date      project_end_date  \
queried_project_id                                               
P50CA070907         1996-09-30T12:09:00Z  2025-08-31T12:08:00Z   

                   opportunity_number api_source_search org_name org_city  \
queried_project_id                                                          
P50CA070907                       NaN               NaN      NaN      NaN   

                   org_state org_country program.program_id  
queried_project_id                                           
P50CA070907              NaN         NaN                NaN

In [41]:
actual = projects_df['project_start_date'].iloc[0] 

expected = test_df[test_df['award_notice_date'] == oldest]['project_start_date'].iloc[0]

actual == expected


True

In [141]:
def test_newest_oldest_project_val(grants_df, project_id):
    """Test that correct project values are pulled from either newest or oldest 
    grants as defined in config."""

    # Isolate a single project from input df
    df = grants_df[grants_df['queried_project_id'] == project_id]

    
    # Build empty projects df
    projects_df = pd.DataFrame(columns=field_aggregation_map.keys())
    
    # Set date column
    date_column = 'award_notice_date'

    # Get oldest and newest dates
    oldest = (df.sort_values(by=date_column, ascending=True)
              [date_column].unique().tolist()[0])
    newest = (df.sort_values(by=date_column, ascending=True)
                [date_column].unique().tolist()[-1])
    
    # If oldest or newest values are NaN, use fiscal year
    if pd.isna(oldest) or pd.isna(newest):
        
        date_column = 'fiscal_year'
        oldest = (df.sort_values(by=date_column, ascending=True)
              [date_column].unique().tolist()[0])
        newest = (df.sort_values(by=date_column, ascending=True)
                [date_column].unique().tolist()[-1])

    # Run project workflow to get actual value
    # Validate and aggregate fields based on aggregation type
    for field_name, aggregation_type in field_aggregation_map.items():
        if aggregation_type in ('newest', 'oldest'):
            projects_df[field_name] = get_newest_or_oldest_value(df, 
                                                                 field_name,
                                                                 #date_field=date_column,
                                                                 newest=aggregation_type == 'newest')
            
    # Define dict for testing 
    test_newest_oldest = {
        'project_start_date': oldest,
        'project_end_date': newest,
        'project_title': newest,
        'abstract_text': newest
    }

    # Build empty list
    valid_list = []

    try:

        # Loop through each column and compare actual vs expected values
        for col_name, date in test_newest_oldest.items():
            actual = projects_df[col_name].iloc[0] 
            expected = df[df[date_column] == date][col_name].iloc[0]
        
            if actual == expected:
                valid = True
            else:
                valid = False
                print(f"Mismatch in {col_name} for {project_id}:\n"
                    f"\tActual:   {actual}\n"
                    f"\tExpected: {expected}")
                
            # Add to running list
            valid_list.append(valid)

        # if all(valid_list):
        #     print(f"All good. Actual values match expected for {project_id}.")
    
    except Exception as error:
        print(f"Issue encountered with {project_id}: {error}\n"
              f"{display(df.head(2))}")
        
    return all(valid_list)


In [142]:
test_newest_oldest_project_val(grants_df, 'P50CA070907')

True

In [143]:
all_results = list()
for project in grants_df['queried_project_id'].unique():
    project_results = test_newest_oldest_project_val(grants_df, project)
    all_results.append(project_results)

if all(all_results) == True: print("All good!")

Mismatch in project_end_date for U01CA069853:
	Actual:   2009-02-28T12:02:00Z
	Expected: 2007-12-16T12:12:00Z


In [144]:
grants_df[grants_df['queried_project_id'] == 'U01CA069853']

grant_id queried_project_id  application_id  fiscal_year  \
879   2U01CA069853-09        U01CA069853         6581582         2003   
1014  5U01CA069853-11        U01CA069853         6879625         2005   
1015  5U01CA069853-12        U01CA069853         7027645         2006   
1016  5U01CA069853-13        U01CA069853         7190568         2007   
1127  6U01CA069853-10        U01CA069853         6736274         2004   
1128  6U01CA069853-14        U01CA069853         7617484         2007   

                                      project_title  \
879   Phase I Clinical Trials of Anti-Cancer Agents   
1014  Phase I Clinical Trials of Anti-Cancer Agents   
1015  Phase I Clinical Trials of Anti-Cancer Agents   
1016  Phase I Clinical Trials of Anti-Cancer Agents   
1127  Phase I Clinical Trials of Anti-Cancer Agents   
1128  Phase I Clinical Trials of Anti-Cancer Agents   

                                          abstract_text  \
879   DESCRIPTION (provided by applicant): This prop...   
1014  DESCRIPTION (provided by applicant): This prop...   
1015  DESCRIPTION (provided by applicant): This prop...   
1016  DESCRIPTION (provided by applicant): This prop...   
1127  DESCRIPTION (provided by applicant): This prop...   
1128  DESCRIPTION (provided by applicant): This prop...   

                                               keywords  \
879   antineoplastics; capillary electrophoresis; cl...   
1014  antineoplastics;capillary electrophoresis;clin...   
1015  antineoplastics;capillary electrophoresis;clin...   
1016  Acute;Address;Antibodies;Antineoplastic Agents...   
1127  antineoplastics;capillary electrophoresis;clin...   
1128  Acute;Address;Adoption;Antibodies;Antineoplast...   

     principal_investigators   program_officers  award_amount  ...  \
879          Eric K Rowinsky    Leeann T Jensen        486569  ...   
1014       Anthony W Tolcher    Leeann T Jensen        503913  ...   
1015       Anthony W Tolcher          Roy S. Wu        497588  ...   
1016        Francis J. Giles  S Percy Percy Ivy        321790  ...   
1127       Anthony W Tolcher    Leeann T Jensen        492039  ...   
1128        Francis J. Giles  S Percy Percy Ivy        179544  ...   

         award_notice_date    project_start_date      project_end_date  \
879                    NaN  1995-07-01T12:07:00Z  2004-02-29T12:02:00Z   
1014  2005-03-02T12:03:00Z  1995-07-01T12:07:00Z  2008-02-29T12:02:00Z   
1015  2006-02-23T12:02:00Z  1995-07-01T12:07:00Z  2008-02-29T12:02:00Z   
1016  2007-03-26T12:03:00Z  1995-07-01T12:07:00Z  2007-12-16T12:12:00Z   
1127  2004-03-29T12:03:00Z  1995-07-01T12:07:00Z  2008-02-29T12:02:00Z   
1128  2008-06-06T12:06:00Z  1995-07-01T12:07:00Z  2009-02-28T12:02:00Z   

     opportunity_number api_source_search  \
879       RFA-CA-02-011    nofo_CA-02-011   
1014      RFA-CA-02-011    nofo_CA-02-011   
1015      RFA-CA-02-011    nofo_CA-02-011   
1016      RFA-CA-02-011    nofo_CA-02-011   
1127      RFA-CA-02-011    nofo_CA-02-011   
1128      RFA-CA-02-011    nofo_CA-02-011   

                                     org_name     org_city org_state  \
879        CANCER THERAPY AND RESEARCH CENTER  SAN ANTONIO        TX   
1014       CANCER THERAPY AND RESEARCH CENTER  SAN ANTONIO        TX   
1015       CANCER THERAPY AND RESEARCH CENTER  SAN ANTONIO        TX   
1016       CANCER THERAPY AND RESEARCH CENTER  SAN ANTONIO        TX   
1127       CANCER THERAPY AND RESEARCH CENTER  SAN ANTONIO        TX   
1128  UNIVERSITY OF TEXAS HLTH SCIENCE CENTER  SAN ANTONIO        TX   

        org_country program.program_id  
879   UNITED STATES              etctn  
1014  UNITED STATES              etctn  
1015  UNITED STATES              etctn  
1016  UNITED STATES              etctn  
1127  UNITED STATES              etctn  
1128  UNITED STATES              etctn  

[6 rows x 21 columns]

In [14]:
grants_df[['queried_project_id','program.program_id']].drop_duplicates().reset_index(drop=True)

queried_project_id program.program_id
0          U24CA233218             accsis
1          UG3CA233229             accsis
2          UG3CA233251             accsis
3          UG3CA233282             accsis
4          UG3CA233314             accsis
..                 ...                ...
698        U01CA230705                tlc
699        U01CA230997                tlc
700        U01CA283931                tlc
701        U01CA283935                tlc
702        U24CA230144                tlc

[703 rows x 2 columns]

## Remove Publications before Projects
Testing space for code to remove any publications with a publication date before the project start date of the associated project.  

In [30]:
# Load intermediate files
project_path = '../'+config.PROJECTS_INTERMED_PATH
publication_path = '../'+config.PUBLICATIONS_INTERMED_PATH
grants_path = '../'+config.GRANTS_INTERMED_PATH

# Load projects data
if os.path.exists(project_path):
    projects_exist = True
    df_projects = pd.read_csv(project_path)
    print(f"Loaded Projects file from {project_path}")
else: 
    projects_exist = False
    print(f"No Projects file found.")

# Load publications data
if os.path.exists(publication_path):
    publications_exist = True
    df_publications = pd.read_csv(publication_path)
    print(f"Loaded Publications file from {publication_path}")
else:
    publications_exist = False
    print(f"No Projects file found.")
    
# Load grants data
if os.path.exists(grants_path):
    grants_exist = True
    df_grants = pd.read_csv(grants_path)
    print(f"Loaded Grants file from {grants_path}")
else: 
    grants_exist = False
    print(f"No Grants file found.")


Loaded Projects file from ../data/01_intermediate/2024-01-04/gathered-2024-01-04/project.csv
Loaded Publications file from ../data/01_intermediate/2024-01-04/gathered-2024-01-04/publication.csv
Loaded Grants file from ../data/01_intermediate/2024-01-04/gathered-2024-01-04/grant.csv


In [8]:
display(df_projects.head())
display(df_publications.head())

project_id                                      project_title  \
0  26114002B91  Development and Expansion of Digital Tools to ...   
1  26120080000  Burkitt Lymphoma Genome Sequencing Project (BL...   
2  26120150000  HIV and Tumor Molecular Characterization Proje...   
3  P20CA242611  2/2 Howard-Georgetown Collaborative Partnershi...   
4  P20CA242617  1/2 Howard-Georgetown Collaborative Partnershi...   

                                       abstract_text    project_start_date  \
0  Earlier this year, NCI launched the Cancer Cen...  2019-09-15T12:09:00Z   
1  Burkitt Lymphoma (BL) patients would benefit f...  2008-09-26T12:09:00Z   
2  HIV and Tumor Molecular Characterization Proje...  2018-09-24T12:09:00Z   
3  Overall Abstract The overarching goal of the N...  2019-09-01T12:09:00Z   
4  Overall Abstract The overarching goal of the N...  2019-09-09T12:09:00Z   

       project_end_date opportunity_number                 api_source_search  \
0  2020-09-14T12:09:00Z                NaN  award_26114002B91019F00132-0-0-1   
1  2018-09-25T12:09:00Z                NaN      award_261200800001E-52-0-113   
2  2022-03-31T12:03:00Z                NaN  award_261201500003I-0-26100065-1   
3  2023-08-31T12:08:00Z         PAR-18-911                   nofo_PAR-18-911   
4  2023-08-31T12:08:00Z         PAR-18-911                   nofo_PAR-18-911   

                           org_name    org_city org_state    org_country  \
0                   ICF MACRO, INC.      Reston        VA  UNITED STATES   
1  LEIDOS BIOMEDICAL RESEARCH, INC.   FREDERICK        MD  UNITED STATES   
2  LEIDOS BIOMEDICAL RESEARCH, INC.   FREDERICK        MD  UNITED STATES   
3             GEORGETOWN UNIVERSITY  WASHINGTON        DC  UNITED STATES   
4                 HOWARD UNIVERSITY  WASHINGTON        DC  UNITED STATES   

  program.program_id  
0                c3i  
1               cgci  
2               cgci  
3              pache  
4              pache

coreproject      pmid                                              title  \
0  P50CA062924  10623651  Immunohistochemical labeling for dpc4 mirrors ...   
1  U10CA031946  10627441  Molecular analysis and clinical outcome of adu...   
2  U24CA076518  10627443  Effect of short-term interferon therapy on the...   
3  P50CA058223  10637239  HER-2/neu amplification in benign breast disea...   
4  P50CA062924  10639144  Distinct genetic profiles in colorectal tumors...   

                                             authors publication_date  \
0  R E Wilentz; G H Su; J L Dai; A B Sparks; P Ar...       2000-01-01   
1  J L Slack; C L Willman; J W Andersen; Y P Li; ...       2000-01-15   
2  S Giralt; R Szydlo; J M Goldman; J Veum-Stone;...       2000-01-15   
3  A Stark; B S Hulka; S Joens; D Novotny; A D Th...       2000-01-01   
4          M Toyota; M Ohe-Toyota; N Ahuja; J P Issa       2000-01-18   

   citation_count  relative_citation_ratio  
0           261.0                     5.00  
1            48.0                     0.88  
2            32.0                     0.70  
3            91.0                     1.98  
4           374.0                     6.67

In [9]:
# OUTLINE
# Get list of PMIDs and associated coreproject numbers
# code placeholder

# Add PMID publication date from df_publication and project start date from df_project
# code placeholder

# Identify any publications with publication date before project start date
# code placeholder

# Remove and report publications before project as csv
# code placeholder

# Return publications df with issues removed
# code placeholder

# Anything else needed

In [24]:
# Merge project start dates with publication data
df_merged = pd.merge(left=df_publications, 
                     right=df_projects[['project_id','project_start_date']], 
                     how='left',left_on='coreproject', right_on='project_id',
                     # Ignore duplicate projects listed for multiple programs
                     ).drop_duplicates()
df_merged

coreproject      pmid  \
0      P50CA062924  10623651   
1      U10CA031946  10627441   
2      U24CA076518  10627443   
3      P50CA058223  10637239   
4      P50CA062924  10639144   
...            ...       ...   
26955  R21CA248122  38129585   
26956  UM1CA186689  38135713   
26957  U54CA231652  38136296   
26958  P50CA228944  38149262   
26959  U54CA137788  38152038   

                                                   title  \
0      Immunohistochemical labeling for dpc4 mirrors ...   
1      Molecular analysis and clinical outcome of adu...   
2      Effect of short-term interferon therapy on the...   
3      HER-2/neu amplification in benign breast disea...   
4      Distinct genetic profiles in colorectal tumors...   
...                                                  ...   
26955  Transcriptional state dynamics lead to heterog...   
26956  NCI10066: a Phase 1/2 study of olaparib in com...   
26957  EWSR1::ATF1 Orchestrates the Clear Cell Sarcom...   
26958  GENERALIZED MATRIX DECOMPOSITION REGRESSION: E...   
26959  Brain white matter microstructural changes in ...   

                                                 authors publication_date  \
0      R E Wilentz; G H Su; J L Dai; A B Sparks; P Ar...       2000-01-01   
1      J L Slack; C L Willman; J W Andersen; Y P Li; ...       2000-01-15   
2      S Giralt; R Szydlo; J M Goldman; J Veum-Stone;...       2000-01-15   
3      A Stark; B S Hulka; S Joens; D Novotny; A D Th...       2000-01-01   
4              M Toyota; M Ohe-Toyota; N Ahuja; J P Issa       2000-01-18   
...                                                  ...              ...   
26955  Antara Biswas; Sarthak Sahoo; Gregory M Riedli...       2023-01-21   
26956  Michael Cecchini; James M Cleary; Yu Shyr; Jos...       2023-01-22   
26957  Benjamin B Ozenberger; Li Li; Emily R Wilson; ...       2023-01-08   
26958  Yue Wang; Ali Shojaie; Timothy Randolph; Parke...       2023-01-30   
26959  Ebenezer Daniel; Frank Deng; Sunita K Patel; M...       2023-01-28   

       citation_count  relative_citation_ratio   project_id  \
0               261.0                     5.00  P50CA062924   
1                48.0                     0.88  U10CA031946   
2                32.0                     0.70  U24CA076518   
3                91.0                     1.98  P50CA058223   
4               374.0                     6.67  P50CA062924   
...               ...                      ...          ...   
26955             NaN                      NaN  R21CA248122   
26956             NaN                      NaN  UM1CA186689   
26957             NaN                      NaN  U54CA231652   
26958             NaN                      NaN  P50CA228944   
26959             NaN                      NaN  U54CA137788   

         project_start_date  
0      2007-08-29T12:08:00Z  
1      2010-05-01T12:05:00Z  
2      1998-03-01T12:03:00Z  
3      2009-08-01T12:08:00Z  
4      2007-08-29T12:08:00Z  
...                     ...  
26955  2020-05-15T12:05:00Z  
26956  2014-07-10T12:07:00Z  
26957  2018-09-14T12:09:00Z  
26958  2019-08-01T12:08:00Z  
26959  2008-09-26T12:09:00Z  

[26637 rows x 9 columns]

In [25]:
# Double-check that all publications are included
len(df_publications) == len(df_merged)

True

In [26]:
df_merged.dtypes

coreproject                 object
pmid                         int64
title                       object
authors                     object
publication_date            object
citation_count             float64
relative_citation_ratio    float64
project_id                  object
project_start_date          object
dtype: object

In [28]:
# Get subset where publication pub date is earlier than project start date
df_early_pubs = df_merged[df_merged['publication_date'] < df_merged['project_start_date']]
df_early_pubs

coreproject      pmid  \
0      P50CA062924  10623651   
1      U10CA031946  10627441   
3      P50CA058223  10637239   
4      P50CA062924  10639144   
5      U10CA031946  10655057   
...            ...       ...   
26880  R01CA271601  38042522   
26884  P50CA070907  38045409   
26916  U24CA274582  38081564   
26918  P50CA070907  38081823   
26947  U54CA274506  38114518   

                                                   title  \
0      Immunohistochemical labeling for dpc4 mirrors ...   
1      Molecular analysis and clinical outcome of adu...   
3      HER-2/neu amplification in benign breast disea...   
4      Distinct genetic profiles in colorectal tumors...   
5      Aberrant CpG-island methylation has non-random...   
...                                                  ...   
26880  The role of ROS in tumor infiltrating immune c...   
26884  MitoTracer facilitates the identification of i...   
26916  Enhanced family history-based algorithms incre...   
26918  Deep learning of cell spatial organizations id...   
26947  Neoadjuvant durvalumab plus radiation versus d...   

                                                 authors publication_date  \
0      R E Wilentz; G H Su; J L Dai; A B Sparks; P Ar...       2000-01-01   
1      J L Slack; C L Willman; J W Andersen; Y P Li; ...       2000-01-15   
3      A Stark; B S Hulka; S Joens; D Novotny; A D Th...       2000-01-01   
4              M Toyota; M Ohe-Toyota; N Ahuja; J P Issa       2000-01-18   
5      J F Costello; M C Frühwald; D J Smiraglia; L J...       2000-02-01   
...                                                  ...              ...   
26880  Rushil Shah; Betul Ibis; Monisha Kashyap; Vass...       2023-01-30   
26884  Xuexin Yu; Jing Hu; Yuhao Tan; Mingyao Pan; Ho...       2023-01-22   
26916  Richard L Bradshaw; Kensaku Kawamoto; Jemar R ...       2023-01-09   
26918  Shidan Wang; Ruichen Rong; Qin Zhou; Donghan M...       2023-01-11   
26947  Nasser K Altorki; Zachary H Walsh; Johannes C ...       2023-01-19   

       citation_count  relative_citation_ratio   project_id  \
0               261.0                     5.00  P50CA062924   
1                48.0                     0.88  U10CA031946   
3                91.0                     1.98  P50CA058223   
4               374.0                     6.67  P50CA062924   
5              1056.0                    18.89  U10CA031946   
...               ...                      ...          ...   
26880             0.0                      NaN  R01CA271601   
26884             0.0                      NaN  P50CA070907   
26916             NaN                      NaN  U24CA274582   
26918             NaN                      NaN  P50CA070907   
26947             NaN                      NaN  U54CA274506   

         project_start_date  
0      2007-08-29T12:08:00Z  
1      2010-05-01T12:05:00Z  
3      2009-08-01T12:08:00Z  
4      2007-08-29T12:08:00Z  
5      2010-05-01T12:05:00Z  
...                     ...  
26880  2023-04-17T12:04:00Z  
26884  2023-05-01T12:05:00Z  
26916  2023-09-01T12:09:00Z  
26918  2023-05-01T12:05:00Z  
26947  2023-09-19T12:09:00Z  

[4350 rows x 9 columns]

### Check issues with df_early_pubs
There are some here that seem like they should be checked further: 
- Some of the early publications around year 2000 have project start dates 10 years later (e.g. PMID 10627441 for U10CA31946)
- Some of the recent publications from 2023 have 2023 project start dates. These may need some leniency


Spot check grants of early projects to see if dates are wonky

In [36]:
df = df_grants[df_grants['queried_project_id'] == 'P50CA062924']
df

grant_id queried_project_id  application_id  fiscal_year  \
2625    2P50CA062924-09        P50CA062924         6466113         2002   
2626    2P50CA062924-14        P50CA062924         7245222         2007   
2627    2P50CA062924-19        P50CA062924         8336982         2012   
2628  2P50CA062924-24A1        P50CA062924         9569170         2018   
2629  3P50CA062924-08S1        P50CA062924         6491995         2001   
2630  3P50CA062924-08S2        P50CA062924         6501328         2001   
2631  3P50CA062924-08S3        P50CA062924         6555313         2002   
2632  3P50CA062924-10S1        P50CA062924         6766414         2003   
2633  3P50CA062924-11S1        P50CA062924         6938078         2004   
2634  3P50CA062924-12S1        P50CA062924         7118436         2005   
2635  3P50CA062924-13S1        P50CA062924         7249964         2006   
2636  3P50CA062924-13S2        P50CA062924         7283506         2006   
2637  3P50CA062924-13S3        P50CA062924         7285071         2006   
2638    4P50CA062924-23        P50CA062924         9134682         2016   
2639    5P50CA062924-07        P50CA062924         6137540         2000   
2640    5P50CA062924-08        P50CA062924         6341978         2001   
2641    5P50CA062924-10        P50CA062924         6631424         2003   
2642    5P50CA062924-11        P50CA062924         6758624         2004   
2643    5P50CA062924-12        P50CA062924         6910012         2005   
2644    5P50CA062924-13        P50CA062924         7118600         2006   
2645    5P50CA062924-15        P50CA062924         7491189         2008   
2646    5P50CA062924-16        P50CA062924         7686217         2009   
2647    5P50CA062924-17        P50CA062924         7874610         2010   
2648    5P50CA062924-18        P50CA062924         8096775         2011   
2649    5P50CA062924-20        P50CA062924         8547744         2013   
2650    5P50CA062924-21        P50CA062924         8727973         2014   
2651    5P50CA062924-22        P50CA062924         8916023         2015   
2652    5P50CA062924-25        P50CA062924         9778705         2019   
2653    5P50CA062924-26        P50CA062924        10005974         2020   
2654    5P50CA062924-27        P50CA062924        10246361         2021   
2655    5P50CA062924-28        P50CA062924        10480035         2022   

                          project_title  \
2625   SPORE in Gastrointestinal Cancer   
2626   SPORE in Gastrointestinal Cancer   
2627   SPORE in Gastrointestinal Cancer   
2628  SPORE in Gastrointestinal Cancers   
2629   SPORE IN GASTROINTESTINAL CANCER   
2630   SPORE IN GASTROINTESTINAL CANCER   
2631   SPORE IN GASTROINTESTINAL CANCER   
2632   SPORE in Gastrointestinal Cancer   
2633   SPORE in Gastrointestinal Cancer   
2634   SPORE in Gastrointestinal Cancer   
2635   SPORE in Gastrointestinal Cancer   
2636   SPORE in Gastrointestinal Cancer   
2637   SPORE in Gastrointestinal Cancer   
2638   SPORE in Gastrointestinal Cancer   
2639   SPORE IN GASTROINTESTINAL CANCER   
2640   SPORE IN GASTROINTESTINAL CANCER   
2641   SPORE in Gastrointestinal Cancer   
2642   SPORE in Gastrointestinal Cancer   
2643   SPORE in Gastrointestinal Cancer   
2644   SPORE in Gastrointestinal Cancer   
2645   SPORE in Gastrointestinal Cancer   
2646   SPORE in Gastrointestinal Cancer   
2647   SPORE in Gastrointestinal Cancer   
2648   SPORE in Gastrointestinal Cancer   
2649   SPORE in Gastrointestinal Cancer   
2650   SPORE in Gastrointestinal Cancer   
2651   SPORE in Gastrointestinal Cancer   
2652  SPORE in Gastrointestinal Cancers   
2653  SPORE in Gastrointestinal Cancers   
2654  SPORE in Gastrointestinal Cancers   
2655  SPORE in Gastrointestinal Cancers   

                                          abstract_text  \
2625  DESCRIPTION (provided by applicant) This appli...   
2626  DESCRIPTION (provided by applicant): This appl...   
2627  DESCRIPTION (provided by applicant): This appl...   
2

Something is indeed wonky with the project start dates. Need to check that they are being applied correctly in the projects module.

In [46]:
field_name = 'project_start_date'
sort_new = True

# Determine the appropriate sorting column based on the field name
if field_name in ('project_start_date', 'project_end_date'):
    sort_col = field_name
else:
    sort_col = 'award_notice_date'  # Default sorting column

# Sort values, group by project, and extract the first value
value = (
    df.sort_values(sort_col, ascending=sort_new)
    .groupby('queried_project_id')[field_name]
    .first()
)

value

queried_project_id
P50CA062924    1993-09-30T12:09:00Z
Name: project_start_date, dtype: object

Date sorting was reversed from expected behavior. Projects module fixed and rerun.

## Rerun with updated projects file

In [47]:
# Load intermediate files
project_path = '../'+config.PROJECTS_INTERMED_PATH
publication_path = '../'+config.PUBLICATIONS_INTERMED_PATH
grants_path = '../'+config.GRANTS_INTERMED_PATH

# Load projects data
if os.path.exists(project_path):
    projects_exist = True
    df_projects = pd.read_csv(project_path)
    print(f"Loaded Projects file from {project_path}")
else: 
    projects_exist = False
    print(f"No Projects file found.")

# Load publications data
if os.path.exists(publication_path):
    publications_exist = True
    df_publications = pd.read_csv(publication_path)
    print(f"Loaded Publications file from {publication_path}")
else:
    publications_exist = False
    print(f"No Projects file found.")
    
# Load grants data
if os.path.exists(grants_path):
    grants_exist = True
    df_grants = pd.read_csv(grants_path)
    print(f"Loaded Grants file from {grants_path}")
else: 
    grants_exist = False
    print(f"No Grants file found.")


Loaded Projects file from ../data/01_intermediate/2024-01-04/gathered-2024-01-04/project.csv
Loaded Publications file from ../data/01_intermediate/2024-01-04/gathered-2024-01-04/publication.csv
Loaded Grants file from ../data/01_intermediate/2024-01-04/gathered-2024-01-04/grant.csv


In [51]:
# Double-check new project sorting
df_projects[df_projects['project_id'] == 'P50CA062924']

project_id                      project_title  \
18  P50CA062924  SPORE in Gastrointestinal Cancers   

                                        abstract_text    project_start_date  \
18  This application for continuation of the Speci...  1993-09-30T12:09:00Z   

        project_end_date                                 opportunity_number  \
18  2024-08-31T12:08:00Z  CA95-021;PAR-00-087;PAR-05-156;PAR-10-003;PAR-...   

    api_source_search                  org_name   org_city org_state  \
18  award_P50CA062924  JOHNS HOPKINS UNIVERSITY  BALTIMORE        MD   

      org_country  
18  UNITED STATES

In [48]:
# Merge project start dates with publication data
df_merged = pd.merge(left=df_publications, 
                     right=df_projects[['project_id','project_start_date']], 
                     how='left',left_on='coreproject', right_on='project_id',
                     # Ignore duplicate projects listed for multiple programs
                     ).drop_duplicates()
df_merged

coreproject      pmid  \
0      P50CA062924  10623651   
1      U10CA031946  10627441   
2      U24CA076518  10627443   
3      P50CA058223  10637239   
4      P50CA062924  10639144   
...            ...       ...   
26632  R21CA248122  38129585   
26633  UM1CA186689  38135713   
26634  U54CA231652  38136296   
26635  P50CA228944  38149262   
26636  U54CA137788  38152038   

                                                   title  \
0      Immunohistochemical labeling for dpc4 mirrors ...   
1      Molecular analysis and clinical outcome of adu...   
2      Effect of short-term interferon therapy on the...   
3      HER-2/neu amplification in benign breast disea...   
4      Distinct genetic profiles in colorectal tumors...   
...                                                  ...   
26632  Transcriptional state dynamics lead to heterog...   
26633  NCI10066: a Phase 1/2 study of olaparib in com...   
26634  EWSR1::ATF1 Orchestrates the Clear Cell Sarcom...   
26635  GENERALIZED MATRIX DECOMPOSITION REGRESSION: E...   
26636  Brain white matter microstructural changes in ...   

                                                 authors publication_date  \
0      R E Wilentz; G H Su; J L Dai; A B Sparks; P Ar...       2000-01-01   
1      J L Slack; C L Willman; J W Andersen; Y P Li; ...       2000-01-15   
2      S Giralt; R Szydlo; J M Goldman; J Veum-Stone;...       2000-01-15   
3      A Stark; B S Hulka; S Joens; D Novotny; A D Th...       2000-01-01   
4              M Toyota; M Ohe-Toyota; N Ahuja; J P Issa       2000-01-18   
...                                                  ...              ...   
26632  Antara Biswas; Sarthak Sahoo; Gregory M Riedli...       2023-01-21   
26633  Michael Cecchini; James M Cleary; Yu Shyr; Jos...       2023-01-22   
26634  Benjamin B Ozenberger; Li Li; Emily R Wilson; ...       2023-01-08   
26635  Yue Wang; Ali Shojaie; Timothy Randolph; Parke...       2023-01-30   
26636  Ebenezer Daniel; Frank Deng; Sunita K Patel; M...       2023-01-28   

       citation_count  relative_citation_ratio   project_id  \
0               261.0                     5.00  P50CA062924   
1                48.0                     0.88  U10CA031946   
2                32.0                     0.70  U24CA076518   
3                91.0                     1.98  P50CA058223   
4               374.0                     6.67  P50CA062924   
...               ...                      ...          ...   
26632             NaN                      NaN  R21CA248122   
26633             NaN                      NaN  UM1CA186689   
26634             NaN                      NaN  U54CA231652   
26635             NaN                      NaN  P50CA228944   
26636             NaN                      NaN  U54CA137788   

         project_start_date  
0      1993-09-30T12:09:00Z  
1      1983-09-30T12:09:00Z  
2      1998-03-01T12:03:00Z  
3      1992-09-30T12:09:00Z  
4      1993-09-30T12:09:00Z  
...                     ...  
26632  2020-05-15T12:05:00Z  
26633  2014-07-10T12:07:00Z  
26634  2018-09-14T12:09:00Z  
26635  2019-08-01T12:08:00Z  
26636  2008-09-26T12:09:00Z  

[26637 rows x 9 columns]

In [100]:
# Get subset where publication pub date is earlier than project start date
df_early_pubs = df_merged[
    # Convert dates to datetime for more accurate comparison
    pd.to_datetime(df_merged['publication_date'],utc=True) < 
    pd.to_datetime(df_merged['project_start_date'],utc=True)]
df_early_pubs

coreproject      pmid  \
199    U54CA118638  11572412   
307    U54CA118638  12088200   
474    P50CA098131  12842082   
520    P50CA098131  14508084   
597    P50CA100632  14763161   
...            ...       ...   
26500  U54CA272691  38001623   
26506  U54CA272691  38007476   
26558  R01CA271601  38042522   
26593  U24CA274582  38081564   
26624  U54CA274506  38114518   

                                                   title  \
199    Lack of adequate attention to elevated blood p...   
307    Inner-city African American women who failed t...   
474    Targeting the TGF beta signaling network in hu...   
520    Inhibiting tyrosine kinases: successes and lim...   
597    The mitogen-activated protein kinase signaling...   
...                                                  ...   
26500  Type 2 Cystatins and Their Roles in the Regula...   
26506  Mechanisms of ferroptosis and targeted therape...   
26558  The role of ROS in tumor infiltrating immune c...   
26593  Enhanced family history-based algorithms incre...   
26624  Neoadjuvant durvalumab plus radiation versus d...   

                                                 authors publication_date  \
199    E O Quaye; E Alema-Mensah; C Omeogu; D Alvarez...       2001-01-01   
307    John F C Sung; Ernest Alema-Mensah; Daniel S B...       2002-01-01   
474                       Nancy Dumont; Carlos L Arteaga       2003-06-01   
520                                     Carlos L Arteaga       2003-01-01   
597    Michele Milella; Steven M Kornblau; Michael An...       2003-06-01   
...                                                  ...              ...   
26500                        Zijun Zhang; Fenghuang Zhan       2023-01-10   
26506  Tiantian Yu; Zijun Y Xu-Monette; Li Yu; Yong L...       2023-01-25   
26558  Rushil Shah; Betul Ibis; Monisha Kashyap; Vass...       2023-01-30   
26593  Richard L Bradshaw; Kensaku Kawamoto; Jemar R ...       2023-01-09   
26624  Nasser K Altorki; Zachary H Walsh; Johannes C ...       2023-01-19   

       citation_count  relative_citation_ratio   project_id  \
199               2.0                     0.04  U54CA118638   
307              28.0                     0.79  U54CA118638   
474             199.0                     3.37  P50CA098131   
520              11.0                     0.19  P50CA098131   
597              19.0                     0.34  P50CA100632   
...               ...                      ...          ...   
26500             0.0                      NaN  U54CA272691   
26506             0.0                      NaN  U54CA272691   
26558             0.0                      NaN  R01CA271601   
26593             NaN                      NaN  U24CA274582   
26624             NaN                      NaN  U54CA274506   

         project_start_date  
199    2005-09-30T12:09:00Z  
307    2005-09-30T12:09:00Z  
474    2003-08-07T12:08:00Z  
520    2003-08-07T12:08:00Z  
597    2003-08-05T12:08:00Z  
...                     ...  
26500  2023-09-01T12:09:00Z  
26506  2023-09-01T12:09:00Z  
26558  2023-04-17T12:04:00Z  
26593  2023-09-01T12:09:00Z  
26624  2023-09-19T12:09:00Z  

[960 rows x 9 columns]

In [103]:
# Consider adding a buffer to only catch egregious mismatches, especially in the
# more recent publications
buffer_time = '365 days'
print(pd.to_datetime(df_early_pubs.iloc[0,8],utc=True))
print(pd.to_datetime(df_early_pubs.iloc[0,8],utc=True) + pd.Timedelta(buffer_time))

2005-09-30 12:09:00+00:00
2006-09-30 12:09:00+00:00


In [109]:
# Set allowance for time oddities
buffer_time = '365 days'

# Get subset where publication pub date is earlier than project start date + 1yr
df_early_pubs = df_merged[
    # Convert dates to datetime for more accurate comparison
    (pd.to_datetime(df_merged['publication_date'],utc=True) 
     + pd.Timedelta(buffer_time))
     < pd.to_datetime(df_merged['project_start_date'],utc=True)]

df_early_pubs

coreproject      pmid  \
199    U54CA118638  11572412   
307    U54CA118638  12088200   
699    U54CA132383  15219164   
700    U54CA118948  15221990   
701    U54CA118623  15221990   
...            ...       ...   
18002  R01CA261723  33415023   
18639  R01CA285308  33720354   
20217  U01CA227550  34541587   
20622  U24CA271012  34719806   
20681  U01CA274304  34746705   

                                                   title  \
199    Lack of adequate attention to elevated blood p...   
307    Inner-city African American women who failed t...   
699        Essential oncology facts for the float nurse.   
700    Postsurgical disparity in survival between Afr...   
701    Postsurgical disparity in survival between Afr...   
...                                                  ...   
18002  Classify multicategory outcome in patients wit...   
18639  Molecular and cellular features of CTLA-4 bloc...   
20217  Computational Horizons In Cancer (CHIC): Devel...   
20622  Mass spectrometry-based targeted proteomics fo...   
20681  Signal processing capacity of the cellular sen...   

                                                 authors publication_date  \
199    E O Quaye; E Alema-Mensah; C Omeogu; D Alvarez...       2001-01-01   
307    John F C Sung; Ernest Alema-Mensah; Daniel S B...       2002-01-01   
699                          Teresa G Leon; Marilyn Pase       2004-06-01   
700    Dominik Alexander; Chakrapani Chatla; Ellen Fu...       2004-07-01   
701    Dominik Alexander; Chakrapani Chatla; Ellen Fu...       2004-07-01   
...                                                  ...              ...   
18002      Fei Deng; Lanlan Shen; He Wang; Lanjing Zhang       2020-01-01   
18639  Livius Penter; Yi Zhang; Alexandra Savell; Ted...       2021-06-10   
20217  G Stamatakos; Dimitra Dionysiou; Fay Misichron...       2014-11-01   
20622  Tai-Tu Lin; Tong Zhang; Reta B Kitata; Tao Liu...       2021-01-31   
20681  Hye-Ran Moon; Soutick Saha; Andrew Mugler; Bum...       2021-01-07   

       citation_count  relative_citation_ratio   project_id  \
199               2.0                     0.04  U54CA118638   
307              28.0                     0.79  U54CA118638   
699               6.0                     0.25  U54CA132383   
700              45.0                     0.96  U54CA118948   
701              45.0                     0.96  U54CA118623   
...               ...                      ...          ...   
18002            20.0                     1.97  R01CA261723   
18639             8.0                     0.83  R01CA285308   
20217            16.0                     1.58  U01CA227550   
20622            23.0                     1.73  U24CA271012   
20681             1.0                     0.39  U01CA274304   

         project_start_date  
199    2005-09-30T12:09:00Z  
307    2005-09-30T12:09:00Z  
699    2007-09-30T12:09:00Z  
700    2005-09-23T12:09:00Z  
701    2005-09-28T12:09:00Z  
...                     ...  
18002  2021-07-01T12:07:00Z  
18639  2023-09-19T12:09:00Z  
20217  2018-06-01T12:06:00Z  
20622  2022-06-01T12:06:00Z  
20681  2022-09-01T12:09:00Z  

[205 rows x 9 columns]

In [110]:
df_grants[df_grants['queried_project_id'] == 'R01CA261723']

grant_id queried_project_id  application_id  fiscal_year  \
1849  1R01CA261723-01        R01CA261723        10272375         2021   
2001  5R01CA261723-02        R01CA261723        10437934         2022   
2002  5R01CA261723-03        R01CA261723        10668346         2023   

                                          project_title  \
1849  Epigenetically Engineered Mouse Model for Lung...   
2001  Epigenetically Engineered Mouse Model for Lung...   
2002  Epigenetically Engineered Mouse Model for Lung...   

                                          abstract_text  \
1849  PROJECT SUMMARY (Abstract) The importance of a...   
2001  PROJECT SUMMARY (Abstract) The importance of a...   
2002  PROJECT SUMMARY (Abstract) The importance of a...   

                                               keywords  \
1849  Aberrant DNA Methylation;Adenocarcinoma;Affect...   
2001  Aberrant DNA Methylation;Adenocarcinoma;Affect...   
2002  Aberrant DNA Methylation;Acceleration;Adenocar...   

     principal_investigators          program_officers  award_amount  ...  \
1849             Lanlan Shen          Joanna M. Watson        365308  ...   
2001             Lanlan Shen          Joanna M. Watson        349064  ...   
2002             Lanlan Shen  Justin Michael Benavidez        353455  ...   

         award_notice_date    project_start_date      project_end_date  \
1849  2021-06-24T12:06:00Z  2021-07-01T12:07:00Z  2026-06-30T12:06:00Z   
2001  2022-07-19T12:07:00Z  2021-07-01T12:07:00Z  2026-06-30T12:06:00Z   
2002  2023-07-03T12:07:00Z  2021-07-01T12:07:00Z  2026-06-30T12:06:00Z   

     opportunity_number api_source_search                    org_name  \
1849         PAR-20-131   nofo_PAR-20-131  BAYLOR COLLEGE OF MEDICINE   
2001         PAR-20-131   nofo_PAR-20-131  BAYLOR COLLEGE OF MEDICINE   
2002         PAR-20-131   nofo_PAR-20-131  BAYLOR COLLEGE OF MEDICINE   

     org_city org_state    org_country program.program_id  
1849  HOUSTON        TX  UNITED STATES                omf  
2001  HOUSTON        TX  UNITED STATES                omf  
2002  HOUSTON        TX  UNITED STATES                omf  

[3 rows x 21 columns]

Does removing these publications make sense? Despite the date differences, spot checks seem to be related.  

For example, PMID 33415023 (published 2020) for project R01CA261723 (start date 2021) are related when reading the [NIH RePORTER Project Abstract](https://reporter.nih.gov/search/gwrXAO7Hm0y1p12dTZwH_g/project-details/10272375) vs the [PubMed Article](https://pubmed.ncbi.nlm.nih.gov/33415023/) which lists the project as a funding source.

In [126]:
# Set allowance for time oddities
buffer_time = '0'

# Get subset where publication pub date is earlier than project start date + 1yr
df_early_pubs = df_merged[
    # Convert dates to datetime for more accurate comparison
    (pd.to_datetime(df_merged['publication_date'],utc=True) 
     + pd.Timedelta(buffer_time))
     < pd.to_datetime(df_merged['project_start_date'],utc=True)]

df_early_pubs

coreproject      pmid  \
199    U54CA118638  11572412   
307    U54CA118638  12088200   
474    P50CA098131  12842082   
520    P50CA098131  14508084   
597    P50CA100632  14763161   
...            ...       ...   
26500  U54CA272691  38001623   
26506  U54CA272691  38007476   
26558  R01CA271601  38042522   
26593  U24CA274582  38081564   
26624  U54CA274506  38114518   

                                                   title  \
199    Lack of adequate attention to elevated blood p...   
307    Inner-city African American women who failed t...   
474    Targeting the TGF beta signaling network in hu...   
520    Inhibiting tyrosine kinases: successes and lim...   
597    The mitogen-activated protein kinase signaling...   
...                                                  ...   
26500  Type 2 Cystatins and Their Roles in the Regula...   
26506  Mechanisms of ferroptosis and targeted therape...   
26558  The role of ROS in tumor infiltrating immune c...   
26593  Enhanced family history-based algorithms incre...   
26624  Neoadjuvant durvalumab plus radiation versus d...   

                                                 authors publication_date  \
199    E O Quaye; E Alema-Mensah; C Omeogu; D Alvarez...       2001-01-01   
307    John F C Sung; Ernest Alema-Mensah; Daniel S B...       2002-01-01   
474                       Nancy Dumont; Carlos L Arteaga       2003-06-01   
520                                     Carlos L Arteaga       2003-01-01   
597    Michele Milella; Steven M Kornblau; Michael An...       2003-06-01   
...                                                  ...              ...   
26500                        Zijun Zhang; Fenghuang Zhan       2023-01-10   
26506  Tiantian Yu; Zijun Y Xu-Monette; Li Yu; Yong L...       2023-01-25   
26558  Rushil Shah; Betul Ibis; Monisha Kashyap; Vass...       2023-01-30   
26593  Richard L Bradshaw; Kensaku Kawamoto; Jemar R ...       2023-01-09   
26624  Nasser K Altorki; Zachary H Walsh; Johannes C ...       2023-01-19   

       citation_count  relative_citation_ratio   project_id  \
199               2.0                     0.04  U54CA118638   
307              28.0                     0.79  U54CA118638   
474             199.0                     3.37  P50CA098131   
520              11.0                     0.19  P50CA098131   
597              19.0                     0.34  P50CA100632   
...               ...                      ...          ...   
26500             0.0                      NaN  U54CA272691   
26506             0.0                      NaN  U54CA272691   
26558             0.0                      NaN  R01CA271601   
26593             NaN                      NaN  U24CA274582   
26624             NaN                      NaN  U54CA274506   

         project_start_date  
199    2005-09-30T12:09:00Z  
307    2005-09-30T12:09:00Z  
474    2003-08-07T12:08:00Z  
520    2003-08-07T12:08:00Z  
597    2003-08-05T12:08:00Z  
...                     ...  
26500  2023-09-01T12:09:00Z  
26506  2023-09-01T12:09:00Z  
26558  2023-04-17T12:04:00Z  
26593  2023-09-01T12:09:00Z  
26624  2023-09-19T12:09:00Z  

[960 rows x 9 columns]

In [144]:
df_early_pubs['day_diff'] = (pd.to_datetime(df_early_pubs['project_start_date'],utc=True) 
                             - pd.to_datetime(df_early_pubs['publication_date'],utc=True)).dt.days
df_early_pubs

C:\Users\dormanza\AppData\Local\Temp\8\ipykernel_12492\3219653963.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_early_pubs['day_diff'] = (pd.to_datetime(df_early_pubs['project_start_date'],utc=True)


coreproject      pmid  \
199    U54CA118638  11572412   
307    U54CA118638  12088200   
474    P50CA098131  12842082   
520    P50CA098131  14508084   
597    P50CA100632  14763161   
...            ...       ...   
26500  U54CA272691  38001623   
26506  U54CA272691  38007476   
26558  R01CA271601  38042522   
26593  U24CA274582  38081564   
26624  U54CA274506  38114518   

                                                   title  \
199    Lack of adequate attention to elevated blood p...   
307    Inner-city African American women who failed t...   
474    Targeting the TGF beta signaling network in hu...   
520    Inhibiting tyrosine kinases: successes and lim...   
597    The mitogen-activated protein kinase signaling...   
...                                                  ...   
26500  Type 2 Cystatins and Their Roles in the Regula...   
26506  Mechanisms of ferroptosis and targeted therape...   
26558  The role of ROS in tumor infiltrating immune c...   
26593  Enhanced family history-based algorithms incre...   
26624  Neoadjuvant durvalumab plus radiation versus d...   

                                                 authors publication_date  \
199    E O Quaye; E Alema-Mensah; C Omeogu; D Alvarez...       2001-01-01   
307    John F C Sung; Ernest Alema-Mensah; Daniel S B...       2002-01-01   
474                       Nancy Dumont; Carlos L Arteaga       2003-06-01   
520                                     Carlos L Arteaga       2003-01-01   
597    Michele Milella; Steven M Kornblau; Michael An...       2003-06-01   
...                                                  ...              ...   
26500                        Zijun Zhang; Fenghuang Zhan       2023-01-10   
26506  Tiantian Yu; Zijun Y Xu-Monette; Li Yu; Yong L...       2023-01-25   
26558  Rushil Shah; Betul Ibis; Monisha Kashyap; Vass...       2023-01-30   
26593  Richard L Bradshaw; Kensaku Kawamoto; Jemar R ...       2023-01-09   
26624  Nasser K Altorki; Zachary H Walsh; Johannes C ...       2023-01-19   

       citation_count  relative_citation_ratio   project_id  \
199               2.0                     0.04  U54CA118638   
307              28.0                     0.79  U54CA118638   
474             199.0                     3.37  P50CA098131   
520              11.0                     0.19  P50CA098131   
597              19.0                     0.34  P50CA100632   
...               ...                      ...          ...   
26500             0.0                      NaN  U54CA272691   
26506             0.0                      NaN  U54CA272691   
26558             0.0                      NaN  R01CA271601   
26593             NaN                      NaN  U24CA274582   
26624             NaN                      NaN  U54CA274506   

         project_start_date  day_diff  
199    2005-09-30T12:09:00Z      1733  
307    2005-09-30T12:09:00Z      1368  
474    2003-08-07T12:08:00Z        67  
520    2003-08-07T12:08:00Z       218  
597    2003-08-05T12:08:00Z        65  
...                     ...       ...  
26500  2023-09-01T12:09:00Z       234  
26506  2023-09-01T12:09:00Z       219  
26558  2023-04-17T12:04:00Z        77  
26593  2023-09-01T12:09:00Z       235  
26624  2023-09-19T12:09:00Z       243  

[960 rows x 10 columns]

In [138]:
df_associated_grants = pd.merge(left=df_early_pubs[['coreproject']],
                                right=df_grants,
                                how='left',
                                left_on='coreproject',right_on='queried_project_id'
                                ).drop('coreproject',axis=1).drop_duplicates()
df_associated_grants

grant_id queried_project_id  application_id  fiscal_year  \
0       2U54CA118638-16        U54CA118638        10327909         2021   
1       5U54CA118638-17        U54CA118638        10491940         2022   
2       5U54CA118638-18        U54CA118638        10693349         2023   
6       1P50CA098131-01        P50CA098131         6558032         2003   
7       2P50CA098131-06        P50CA098131         7504833         2008   
...                 ...                ...             ...          ...   
6851  1U01CA274576-01A1        U01CA274576        10642562         2023   
6853    1U54CA283762-01        U54CA283762        10733310         2023   
6855  1U54CA272691-01A1        U54CA272691        10745010         2023   
6858  1U24CA274582-01A1        U24CA274582        10741231         2023   
6859  1U54CA274506-01A1        U54CA274506        10729383         2023   

                                          project_title  \
0     1/3 Morehouse School of Medicine/Tuskegee Univ...   
1     1/3 Morehouse School of Medicine/Tuskegee Univ...   
2     1/3 Morehouse School of Medicine/Tuskegee Univ...   
6                                SPORE in Breast Cancer   
7                                SPORE in Breast Cancer   
...                                                 ...   
6851  Robust privacy preserving distributed analysis...   
6853  United for Health Equity - Living PDX Program ...   
6855  Cancer Prevention-Interception Against MGUS Pr...   
6858  GARDE: Scalable Clinical Decision Support for ...   
6859      Center for Cancer Systems Therapeutics (CaST)   

                                          abstract_text  \
0     OVERALL: PROJECT SUMMARY: (Overall is identica...   
1     OVERALL: PROJECT SUMMARY: (Overall is identica...   
2     OVERALL: PROJECT SUMMARY: (Overall is identica...   
6     This Breast Cancer SPORE application is being ...   
7     DESCRIPTION (provided by applicant): This is t...   
...                                                 ...   
6851  Project Summary Privacy-preserving distributed...   
6853  PROJECT SUMMARY/ABSTRACT – OVERALL The United ...   
6855  Project Summary Monoclonal gammopathy of undet...   
6858  PROJECT SUMMARY/ABSTRACT Evidence supports ind...   
6859  PROJECT SUMMARY Patients with aggressive cance...   

                                               keywords  \
0     Address;Adherence;African American;Alabama;App...   
1     Address;Adherence;African American;African Ame...   
2     Acceleration;Address;Adherence;African America...   
6     biomarker; breast neoplasms; cancer risk; clin...   
7                                                   NaN   
...                                                 ...   
6851  Address;Architecture;Area;Biomedical Research;...   
6853  Address;Advanced Malignant Neoplasm;Affect;Awa...   
6855  Affect;Age Years;American;Amyloid;Automobile D...   
6858  Academic Medical Centers;Address;Adoption;Adva...   
6859  Acceleration;Address;Advanced Malignant Neopla...   

                            principal_investigators  \
0     Upender Manne; Brian M. Rivers; Clayton Yates   
1     Upender Manne; Brian M. Rivers; Clayton Yates   
2     Upender Manne; Brian M. Rivers; Clayton Yates   
6                                  Carlos L Arteaga   
7                                  Carlos L Arteaga   
...                                             ...   
6851                        Xiaoqian Jiang; Qi Long   
6853         Joshua (Chuck) Harrell; Jose G Trevino   
6855                                        Yong Li   
6858           Guilherme Del Fiol; Kensaku Kawamoto   
6859                 Andrea Califano; Barry H Honig   

                      program_officers  award_amount  ...  \
0              Laritza Maria Rodriguez       1178009  ...   
1              Laritza Maria Rodriguez       1142009  ...   
2              Laritza Maria Rodriguez       1152375  ...   
6                      Joyann Courtney       2500000  ...   
7                    

In [140]:
df_early_pubs.coreproject.nunique() == df_associated_grants.queried_project_id.nunique()

True

In [142]:
df_early_pubs.to_csv('removedEarlyPublicationsReport.csv', index=False)
df_associated_grants.to_csv('removedEarlyPublicationsAssociatedGrants.csv', index=False)

In [143]:
df_publications

coreproject      pmid  \
0      P50CA062924  10623651   
1      U10CA031946  10627441   
2      U24CA076518  10627443   
3      P50CA058223  10637239   
4      P50CA062924  10639144   
...            ...       ...   
26632  R21CA248122  38129585   
26633  UM1CA186689  38135713   
26634  U54CA231652  38136296   
26635  P50CA228944  38149262   
26636  U54CA137788  38152038   

                                                   title  \
0      Immunohistochemical labeling for dpc4 mirrors ...   
1      Molecular analysis and clinical outcome of adu...   
2      Effect of short-term interferon therapy on the...   
3      HER-2/neu amplification in benign breast disea...   
4      Distinct genetic profiles in colorectal tumors...   
...                                                  ...   
26632  Transcriptional state dynamics lead to heterog...   
26633  NCI10066: a Phase 1/2 study of olaparib in com...   
26634  EWSR1::ATF1 Orchestrates the Clear Cell Sarcom...   
26635  GENERALIZED MATRIX DECOMPOSITION REGRESSION: E...   
26636  Brain white matter microstructural changes in ...   

                                                 authors publication_date  \
0      R E Wilentz; G H Su; J L Dai; A B Sparks; P Ar...       2000-01-01   
1      J L Slack; C L Willman; J W Andersen; Y P Li; ...       2000-01-15   
2      S Giralt; R Szydlo; J M Goldman; J Veum-Stone;...       2000-01-15   
3      A Stark; B S Hulka; S Joens; D Novotny; A D Th...       2000-01-01   
4              M Toyota; M Ohe-Toyota; N Ahuja; J P Issa       2000-01-18   
...                                                  ...              ...   
26632  Antara Biswas; Sarthak Sahoo; Gregory M Riedli...       2023-01-21   
26633  Michael Cecchini; James M Cleary; Yu Shyr; Jos...       2023-01-22   
26634  Benjamin B Ozenberger; Li Li; Emily R Wilson; ...       2023-01-08   
26635  Yue Wang; Ali Shojaie; Timothy Randolph; Parke...       2023-01-30   
26636  Ebenezer Daniel; Frank Deng; Sunita K Patel; M...       2023-01-28   

       citation_count  relative_citation_ratio  
0               261.0                     5.00  
1                48.0                     0.88  
2                32.0                     0.70  
3                91.0                     1.98  
4               374.0                     6.67  
...               ...                      ...  
26632             NaN                      NaN  
26633             NaN                      NaN  
26634             NaN                      NaN  
26635             NaN                      NaN  
26636             NaN                      NaN  

[26637 rows x 7 columns]